state/execution.go

```diff
@@ -128,16 +132,13 @@ func (blockExec *BlockExecutor) ApplyBlock(
 	state State, blockID types.BlockID, block *types.Block,
 ) (State, int64, error) {
 
-	if err := blockExec.ValidateBlock(state, block); err != nil {
+	if err := validateBlock(state, block); err != nil {
```

Comment:
Why do we validate block twice. Once when we are voting on it and then again when we are applying it. Isn't this part redundant. And for the fast sync can't we just call `ValidateBlock()` before applying the state change

Reply:
there's an issue for this: https://github.com/tendermint/tendermint/issues/1032. 😄 We could reduce it and do a cache like the issue suggests.

Reply:
lol I'm impressed with your memory of all these issues. But what would be caching? the BlockID? why would we need to cache everything. This block should already have been validated